### ISSUE TITLE
collectLogs: false doesn't exclude pipeline logs collection

### ISSUE THREAD

**Issue by pangorgo** [jenkinsci/datadog-plugin#209]
**Describe the bug**
Not sure if this is the bug or I misunderstood plugin configuration. What I want to achieve is to collect logs from all but one Jenkins pipeline.

To do so I've added to all pipelines following:
```
  options {
    ...
    datadog(collectLogs: true)
    ...
  }
```

This single pipeline that I would like to exclude from logs collection has `collectLogs` set to false:
```
  options {
    ...
    datadog(collectLogs: false)
    ...
  }
```

Global "Enable Log Collection" is set to `true`. When I set it to `false` then no logs are collected at all regardless of `collectLogs` pipeline configuration.

Despite `datadog(collectLogs: false)`, logs are still collected from this pipeline. 

**Expected behavior**
I expected that logs would not be collected from pipeline where `collectLogs: false` is configured.

**Screenshots**
![image](https://user-images.githubusercontent.com/5236150/119844966-2c133800-bf09-11eb-8c62-50800da79b85.png)

**Environment and Versions (please complete the following information):**
A clear and precise description of your setup:
  - datadog plugin version: `2.9.0`
  - Jenkins version: `2.277.1`.

**Comment by FlorianVeaux** [comment]
Hello,
The expected configuration is indeed to have `Global Log Collection` disabled, and then enable log collection per pipeline.
Do you see any relevant logs either in your Job or in the Jenkins System Log ?

**Comment by pangorgo** [comment]
After setting `Global Log Collection` to disabled, logs from jobs with `collectLogs: true` lands in Datadog 🤔 
I am confused because I've tried it before and it was not working.

The difference is that the number of logs is a bit smaller. It seems like some pipeline "meta logs" are not collected. For example:
- who run the job
- checkout logs

Is that expected? Is `Global Log Collection` and `collectLogs: true` somehow treats some logs differently?

Regarding my initial statement that enabling global logs collection to `false` disable logs everywhere - I need to do some more testing to prove that because now it behaves differently.

**Comment by FlorianVeaux** [comment]
> Is that expected? Is Global Log Collection and collectLogs: true somehow treats some logs differently?
Unfortunately yes, we use different parts of the plugin API for log collection. A `ConsoleLogFilter` when `Global Log Collection` is enabled (but we can't control per-pipeline with a ConsoleLogFilter), and a `TaskListenerDecorator` when `collectLogs: true` is set (but the "meta logs" as you say are not exposed to it).
I agree that's a technical limitation.

If you need all "meta logs", what you can do is enable `Global Log Collection` and then add an "exclusion filter" from the Datadog UI to exclude some builds: https://docs.datadoghq.com/logs/guide/getting-started-lwl/#3-create-a-log-pattern-exclusion-filter

**Comment by pangorgo** [comment]
I've done more testing and logs collection after toggling `Global Log Collection` seems to be non deterministic. When I turn it off, sometimes logs are stopped being collected right away. Other time when I turn it off, for some time aforementioned "meta logs" stop being collected and then all logs are stopped being collected.

**Comment by FlorianVeaux** [comment]
Indeed, what we do in the plugin is that log collection for a build is determined when the build starts. So any build started before any modification of the `Global Log Collection` option will not take the modification into account.

Does that create any issue for you? How important would you consider that? This is something we can possibly work on

 For `meta logs` I'm not really sure, that shouldn't make any difference and I wasn't able to reproduce.

**Comment by pangorgo** [comment]
> So any build started before any modification of the Global Log Collection option will not take the modification into account

That would be a desired behaviour and I expect it to work like that. The problem is that sometimes also jobs started after switching are not affected by setting the option, or are affected after some time (somewhere between minutes and one hour).

I wonder how can I help you to reproduce that, because also I don't see anything interesting in Jenkins logs. Do you have any suggestions where can I do? Can I somehow change log level for plugin, or see plugin logs in isolation?

**Comment by FlorianVeaux** [comment]
To be fair we don't really control that part. We just hook when Jenkins tells us and we use the current plugin configuration to decide whether or not to forward all the logs to Datadog: https://github.com/jenkinsci/datadog-plugin/blob/master/src/main/java/org/datadog/jenkins/plugins/datadog/logs/DatadogConsoleLogFilter.java#L55
There is no additional caching on our side so I'm not too sure we can have any control on that behavior.

Do you find any pattern worth noticing? i.e is it only for pipeline jobs and not freestyle jobs? Or maybe only the jobs running on a specific jenkins agent?

Also you mention that some jobs are not affected by setting the option, are they not affected *at all* or are they affected after some time and/or a restart of the jenkins process?

**Comment by pangorgo** [comment]
@FlorianVeaux I will do some more testing to find any patters and will come back to you if I find anything helpful.

**Comment by pangorgo** [comment]
I've only managed to find such log records. They are not followed by any stacktrace:

```
2021-06-09 17:18:27.688+0000 [id=175588]        SEVERE  o.d.j.p.datadog.DatadogUtilities#severe: java.util.logging.ErrorManager: 2
2021-06-09 17:18:27.688+0000 [id=175588]        INFO    o.d.j.p.d.c.DogStatsDClient#reinitializeLogger: Re/Initialize Datadog-Plugin Logger: hostname = localhost, logCollectionPort = 8125

2021-06-09 17:21:06.848+0000 [id=174818]        SEVERE  o.d.j.p.datadog.DatadogUtilities#severe: java.util.logging.ErrorManager: 2
2021-06-09 17:21:06.849+0000 [id=174818]        INFO    o.d.j.p.d.c.DogStatsDClient#reinitializeLogger: Re/Initialize Datadog-Plugin Logger: hostname = localhost, logCollectionPort = 8125

2021-06-09 17:25:33.464+0000 [id=174818]        SEVERE  o.d.j.p.datadog.DatadogUtilities#severe: java.util.logging.ErrorManager: 2
2021-06-09 17:25:33.465+0000 [id=174818]        INFO    o.d.j.p.d.c.DogStatsDClient#reinitializeLogger: Re/Initialize Datadog-Plugin Logger: hostname = localhost, logCollectionPort = 8125

2021-06-09 17:29:46.168+0000 [id=176287]        SEVERE  o.d.j.p.datadog.DatadogUtilities#severe: java.util.logging.ErrorManager: 2
2021-06-09 17:29:46.168+0000 [id=176287]        INFO    o.d.j.p.d.c.DogStatsDClient#reinitializeLogger: Re/Initialize Datadog-Plugin Logger: hostname = localhost, logCollectionPort = 8125

2021-06-09 17:32:39.034+0000 [id=176231]        SEVERE  o.d.j.p.datadog.DatadogUtilities#severe: java.util.logging.ErrorManager: 2
2021-06-09 17:32:39.035+0000 [id=176231]        INFO    o.d.j.p.d.c.DogStatsDClient#reinitializeLogger: Re/Initialize Datadog-Plugin Logger: hostname = localhost, logCollectionPort = 8125

2021-06-09 20:56:00.683+0000 [id=192272]        SEVERE  o.d.j.p.datadog.DatadogUtilities#severe: java.util.logging.ErrorManager: 2
2021-06-09 20:56:00.684+0000 [id=192272]        INFO    o.d.j.p.d.c.DogStatsDClient#reinitializeLogger: Re/Initialize Datadog-Plugin Logger: hostname = localhost, logCollectionPort = 8125

2021-06-09 20:57:50.656+0000 [id=192336]        SEVERE  o.d.j.p.datadog.DatadogUtilities#severe: java.util.logging.ErrorManager: 2
2021-06-09 20:57:50.656+0000 [id=192336]        INFO    o.d.j.p.d.c.DogStatsDClient#reinitializeLogger: Re/Initialize Datadog-Plugin Logger: hostname = localhost, logCollectionPort = 8125
```

**Comment by FlorianVeaux** [comment]
Hey, coming back to this. The errors you've copied here seems intriguing but outside of the scope of the initial issue I believe.
Looks like you're using a Datadog agent to forward logs to Datadog, but the plugin is unable to connect to `localhost:8125` for sending logs.

**Comment by pangorgo** [comment]
Hi @FlorianVeaux just a quick update. I've updated both Jenkins and Datadog plugin and problem still exists. Any hints how can I debug it?

**Comment by FlorianVeaux** [comment]
Hi,
You should try to find some patterns. Pipeline jobs and/or Freestyle jobs? running where? how long does it take between changing the option and the change being reflected in-app? etc. Unless we can pinpoint an actual reproducible error it'll be hard to understand exactly what issue you are facing and if it's remediable from the plugin side.

**Comment by hithwen** [comment]
Hi @pangorgo is there any progress on your side with this?

**Comment by ricardojdsilva87** [comment]
Hello @pangorgo sorry for contacting on a closed ticket. Have you found any solution for the SEVERE error logs?
I've tried everything but it seems that I only get the same SEVERE messages but no logs shipped to datadog.
https://github.com/jenkinsci/datadog-plugin/issues/295

I've also added a custom logger on datadog helm chart like described [here](https://docs.datadoghq.com/continuous_integration/setup_pipelines/jenkins/?tab=usingui#enable-job-log-collection)
When trying to select a different port I get the error that the port is closed but the agent status is ok and connected to the local port of the agent.

Thanks
